### PR TITLE
[Feature] Set CI/CD for CapsulaHub repository

### PR DESCRIPTION
**Pull request by MikeSquall**
Related to https://github.com/capsulajs/capsulajs/issues/29

Release notes:
Setup for the CI CD pipeline with deploy on S3 bucket and publish on npm

### CODE REVIEW

File: scripts/deploy_comment.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    return $?
+}
+
+comment(){
+    SERVICE_LINK="[**$SERVICE bundle**](${URL}index.js)"
+    COMMENTS_URL="https://api.github.com/repos/$TRAVIS_REPO_SLUG/issues/$TRAVIS_PULL_REQUEST/comments"
+    COMMENT_TEXT="**Travis-CI** has deployed $SERVICE_LINK"
+
+    [[ "$HAS_CONFIG" == "true" ]] \
+        && CONFIG_LINK="[**$SERVICE configuration**](${URL}configuration/index.json)" \
```
Comment (stephanebenayoun) [inline scripts/deploy_comment.sh]:
```suggestion
        && CONFIG_LINK="[**$SERVICE configuration**](${URL}configuration/)" \
```

File: scripts/deploy_comment.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        && COMMENT_TEXT+=" and $CONFIG_LINK" \
+        && echo "included configuration"
+
+    [[ "$HAS_DOCS" == "true" ]] \
+        && DOC_LINK="[**$SERVICE documentation**](${URL}doc/index.html)" \
+        && COMMENT_TEXT+=" and $DOC_LINK" \
+        && echo "included documentation"
+
+    [[ "$HAS_EXAMPLE" == "true" ]] \
+        && EXAMPLE_LINK="[**$SERVICE documentation**](${URL}example/index.html)" \
```
Comment (stephanebenayoun) [inline scripts/deploy_comment.sh]:
```suggestion
        && EXAMPLE_LINK="[**$SERVICE documentation**](${URL}example/)" \
```

File: scripts/deploy_comment.sh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    COMMENTS_URL="https://api.github.com/repos/$TRAVIS_REPO_SLUG/issues/$TRAVIS_PULL_REQUEST/comments"
+    COMMENT_TEXT="**Travis-CI** has deployed $SERVICE_LINK"
+
+    [[ "$HAS_CONFIG" == "true" ]] \
+        && CONFIG_LINK="[**$SERVICE configuration**](${URL}configuration/index.json)" \
+        && COMMENT_TEXT+=" and $CONFIG_LINK" \
+        && echo "included configuration"
+
+    [[ "$HAS_DOCS" == "true" ]] \
+        && DOC_LINK="[**$SERVICE documentation**](${URL}doc/index.html)" \
```
Comment (stephanebenayoun) [inline scripts/deploy_comment.sh]:
```suggestion
        && DOC_LINK="[**$SERVICE documentation**](${URL}doc/)" \
```

**Comment by stephanebenayoun** [comment]
Also update templates packages with command to publish/deploy (including doc, config, example)

File: .travis.yml [hunk 4]
```diff
@@ -13,6 +13,15 @@ branches:
   only:
     - /.*/
 
+# Keep the npm cache around to speed up installs
+cache:
```
Comment (dmitriyslaym) [inline .travis.yml]: I remember that I has some issues with installing cypress, when I used cache, but if your CI has passed, it seems to be good.

File: .travis.yml [hunk 5]
```diff
@@ -13,6 +13,15 @@ branches:
...
+  npm: true
+  directories:
+    - '$HOME/.npm'
+
+before_install:
+  - echo "//registry.npmjs.org/:_authToken=\${NPM_TOKEN}" > ~/.npmrc
```
Comment (dmitriyslaym) [inline .travis.yml]: Are you saving a token in a file in this step?

File: README.md [hunk 6]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
 | [ui](packages/ui)                       | Common reusable ui components.                                                                                                                                 |
 | [utils](packages/utils)                 | Common reusable helpers, consts, types.                                                                                                                        |
 | [cdn-emulator](packages/cdn-emulator)   | Util package, that includes separate files, that will be served on localhost to imitate cdn links.                                                             |
+
+### CI/CD
```
Comment (dmitriyslaym) [inline README.md]:
```suggestion
## CI/CD
```

File: README.md [hunk 7]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
...
+
+The CI/CD will perform: 
+* deployment on [S3 Bucket](https://capsulajs.s3.amazonaws.com/) 
+* publishing on [npm registry](https://www.npmjs.com/org/capsulajs)
+
+#### Deploy
```
Comment (dmitriyslaym) [inline README.md]:
```suggestion
### Deploy
```

File: README.md [hunk 8]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
...
+In order to deploy to S3, you need to add this npm script to the `package.json`
+
+    "deploy": "SERVICE=$(echo $npm_package_name | cut -d '/' -f 2) && ../../scripts/deploy.sh -s $SERVICE"
+It will deploy all that is in the `dist/` folder which is built during the CI.
```
Comment (dmitriyslaym) [inline README.md]:
```suggestion
It will deploy all the content from `dist/` folder which is built during the CI.  
```

File: README.md [hunk 9]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
...
+#### Deploy
+In order to deploy to S3, you need to add this npm script to the `package.json`
+
+    "deploy": "SERVICE=$(echo $npm_package_name | cut -d '/' -f 2) && ../../scripts/deploy.sh -s $SERVICE"
+It will deploy all that is in the `dist/` folder which is built during the CI.  
+If you want to add some specific directories in it (`doc/`, `configuration/`, `example/`, ...), 
+you can add those npm scripts to the `package.json`
+
+    "build:dist": "<your command> && yarn dist:configuration && yarn dist:doc && yarn dist:example",
+    "dist:configuration": "cpy --parents configuration/ dist/",
```
Comment (dmitriyslaym) [inline README.md]: specify that the package for **cpy** should be installed

File: README.md [hunk 10]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
...
+    "dist:doc": "cpy --parents doc/ dist/",
+    "dist:example": "cpy --parents example/ dist/"
```
Comment (dmitriyslaym) [inline README.md]: is there any difference if we use doc/ with slash at the end or example without slash?

File: README.md [hunk 11]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
...
+
+#### Deploy
+In order to deploy to S3, you need to add this npm script to the `package.json`
+
+    "deploy": "SERVICE=$(echo $npm_package_name | cut -d '/' -f 2) && ../../scripts/deploy.sh -s $SERVICE"
+It will deploy all that is in the `dist/` folder which is built during the CI.  
+If you want to add some specific directories in it (`doc/`, `configuration/`, `example/`, ...), 
+you can add those npm scripts to the `package.json`
+
+    "build:dist": "<your command> && yarn dist:configuration && yarn dist:doc && yarn dist:example",
```
Comment (dmitriyslaym) [inline README.md]: Does it mean that the script will deploy only from dist and we can include different folders inside the dist?

File: README.md [hunk 12]
```diff
@@ -59,3 +59,34 @@ Additional packages, that provide useful common utilities.
...
+The npm publishing will be performed automatically during the CI/CD process.  
+However, you need to **manually publish it for the first time**.  
+If you don't, you will get this error: `lerna ERR! E402 You must sign up for private packages` 
+because of the scoped package name.
+
+If you want, you can enable comments after publishing on npm by adding this npm script to your `package.json`
+
+    "publish:comment": "bash ../../scripts/publish_comment.sh $(echo $npm_package_name)"
+In this case, a bot will comment after publishing the package and will update this comment each time 
+a new package is built for your PR.
```
Comment (dmitriyslaym) [inline README.md]: will the version be updated automatically? What if a user wants to have not a minor update?

File: package.json [hunk 13]
```diff
@@ -10,12 +10,12 @@
   "scripts": {
     "bootstrap": "lerna bootstrap --npm-client=yarn",
     "build": "lerna run build && node ./afterBuild.js --stream",
-    "doc": "lerna run doc --stream",
+    "doc": "lerna run doc --stream --parallel",
```
Comment (dmitriyslaym) [inline package.json]: have you checked - is it really faster?

File: packages/utils/package.json [hunk 14]
```diff
@@ -1,20 +1,22 @@
 {
   "name": "@capsulajs/capsulahub-utils",
   "version": "0.0.1",
-  "description": "Different common utils for Calsulahub",
+  "description": "Different common utils for Capsulahub",
```
Comment (dmitriyslaym) [inline packages/utils/package.json]: my bad(

File: README.md (hunk 11 above)
Comment (MikeSquall) [reply README.md]: yes

File: README.md (hunk 9 above)
Comment (MikeSquall) [reply README.md]: actually, `cpy-cli` is a common dependency so it shouldn't be a problem

File: README.md (hunk 10 above)
Comment (MikeSquall) [reply README.md]: No, but I think this is better to specify the `/` when talking about a directory

File: package.json (hunk 13 above)
Comment (MikeSquall) [reply package.json]: yes
on my local machine, there are 10 seconds less when command is paralleled

File: packages/utils/package.json (hunk 14 above)
Comment (MikeSquall) [reply packages/utils/package.json]: No problem :laughing:

**Pull request merged**